src/wp-includes/class-wp-script-modules.php

```diff
@@ -363,4 +366,119 @@ private function get_src( string $id ): string {
+			 *       $data['script-needs-this-data'] = 'ok';
+			 *       return $data;
+			 *     }
+			 *   );
+			 *
+			 * If the filter returns no data (an empty array), nothing will be embedded in the page.
+			 *
+			 * The data for a given Script Module, if provided, will be JSON serialized in a script
+			 * tag with an ID of the form `wp-scriptmodule-data_{$module_id}`.
+			 *
+			 * The data can be read on the client with a pattern like this:
+			 *
+			 * @example
+			 *   const dataContainer = document.getElementById( 'wp-scriptmodule-data_MyScriptModuleID' );
+			 *   let data = {};
+			 *   if ( dataContainer ) {
+			 *     try {
+			 *       data = JSON.parse( dataContainer.textContent );
+			 *     } catch {}
+			 *   }
+			 *   initMyScriptModuleWithData( data );
+			 *
+			 * @since 6.6.0
+			 *
+			 * @param array $data The data that should be associated with the array.
+			 */
+			$data = apply_filters( "scriptmoduledata_{$module_id}", array() );
+
+			if ( is_array( $data ) && ! empty( $data ) ) {
+				/*
+				 * This data will be printed as JSON inside a script tag like this:
+				 *   <script type="application/json"></script>
+				 *
+				 * A script tag must be closed by a sequence beginning with `</`. It's impossible to
+				 * close a script tag without using `<`. We ensure that `<` is escaped and `/` can
+				 * remain unescaped, so `</script>` will be printed as `\u003C/script\u00E3`.
+				 *
+				 *   - JSON_HEX_TAG: All < and > are converted to \u003C and \u003E.
+				 *   - JSON_UNESCAPED_SLASHES: Don't escape /.
+				 *
+				 * If the page will use UTF-8 encoding, it's safe to print unescaped unicode:
+				 *
+				 *   - JSON_UNESCAPED_UNICODE: Encode multibyte Unicode characters literally (instead of as `\uXXXX`).
+				 *   - JSON_UNESCAPED_LINE_TERMINATORS: The line terminators are kept unescaped when
+				 *     JSON_UNESCAPED_UNICODE is supplied. It uses the same behaviour as it was
+				 *     before PHP 7.1 without this constant. Available as of PHP 7.1.0.
+				 *
+				 * The JSON specification requires encoding in UTF-8, so if the generated HTML page
+				 * is not encoded in UTF-8 then it's not safe to include those literals. They must
+				 * be escaped to avoid encoding issues.
+				 *
+				 * @see https://www.rfc-editor.org/rfc/rfc8259.html for details on encoding requirements.
+				 * @see https://www.php.net/manual/en/json.constants.php for details on these constants.
+				 * @see https://html.spec.whatwg.org/#script-data-state for details on script tag parsing.
+				 */
+				$json_encode_flags = JSON_HEX_TAG | JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_LINE_TERMINATORS;
+				if ( ! is_utf8_charset() ) {
+					$json_encode_flags = JSON_HEX_TAG | JSON_UNESCAPED_SLASHES;
+				}
+
+				wp_print_inline_script_tag(
+					wp_json_encode(
+						$data,
+						$json_encode_flags
+					),
+					array(
+						'type' => 'application/json',
+						'id'   => "wp-scriptmodule-data_{$module_id}",
```

Comment:
The underscore here feels a bit off. Could it be a hyphen (`-`) instead?

Reply:
I'm also noticing some inconsistencies in the naming:

`scriptmodule-data`

Better illustrated inn #6683:

`scriptmoduledata`

<img width="1160" alt="Screenshot 2024-05-31 at 08 45 43" src="https://github.com/WordPress/wordpress-develop/assets/699132/2b1bc613-2957-4698-b884-2728104134f0">

In other places we have function names:

`script_module_data`

<img width="874" alt="Screenshot 2024-05-31 at 08 46 26" src="https://github.com/WordPress/wordpress-develop/assets/699132/ef29140c-4488-438e-8cb4-031b9339e0bb">

It would be helpful to unify it to `script-module-data` and `script_module_data`, respectively, to the conventions required.

Reply:
The underscore was intended to differentiate the static prefix from the dynamic module name. My preference would be to match the filter name:

```suggestion
						'id'   => "scriptmoduledata_{$module_id}",
```

Would that be acceptable?

Reply:
My question was whether the signature of the filter could become easier to read:

`script_module_data_xyz`

Reply:
Done in 7bd30aea3356adaee44743d80b37f410a83450bb / https://github.com/WordPress/gutenberg/pull/62170.

Shall the script tag ID become `script_module_data_{$module_id}` as well?

Reply:
Typically hyphens are used in WordPress for such IDs. 

https://github.com/WordPress/wordpress-develop/blob/23034419b6198cff3398efee32da8c5115bf9517/src/wp-includes/class-wp-scripts.php#L229

https://github.com/WordPress/wordpress-develop/blob/23034419b6198cff3398efee32da8c5115bf9517/src/wp-includes/class-wp-scripts.php#L344

https://github.com/WordPress/wordpress-develop/blob/23034419b6198cff3398efee32da8c5115bf9517/src/wp-includes/class-wp-scripts.php#L229

https://github.com/WordPress/wordpress-develop/blob/23034419b6198cff3398efee32da8c5115bf9517/src/wp-includes/class-wp-scripts.php#L420

https://github.com/WordPress/wordpress-develop/blob/23034419b6198cff3398efee32da8c5115bf9517/src/wp-includes/class-wp-scripts.php#L544

Reply:
The filter and the ID have both been updated:

- filter: `"script_module_data_{$module_id}"`
- ID: `"wp-script-module-data-{$module_id}"`